### ISSUE TITLE
Python 3.6 Spack post_install sbang.py gets UnicodeDecodeError

### ISSUE THREAD

**Issue by lee218llnl** [spack/spack#13489]
When installing perl with Python 3.6, Spack gets a UnicodeDecodeError:

==> Error: UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 3792: ordinal not in range(128)
 
/usr/WS2/gyllen/lee218/thisisaverylongdirectorynametotryandbreaksbanglimitinspack/spack/lib/spack/spack/package.py:1718, in build_process:
       1715                    echo = logger.echo
       1716                    self.log()
       1717
  >>   1718                # Run post install hooks before build stage is removed.
       1719                spack.hooks.post_install(self.spec)
       1720
       1721            # Stop timer.
 
This happens in filter_shebang in lib/spack/spack/hooks/sbang.py on the original_file.read(). This only happens when the path is long enough to trigger this post_install step. It also only fails with Python <=3.6. It does not fail with Python 2.X or Python >= 3.7.

**Comment by alalazo** [comment]
I don't understand at the moment why Python 3.6 is failing with:
```console
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe3 in position 3792: ordinal not in range(128)
```
since, if I am not missing something, `utf-8` should be the default encoding in Python 3. Do you have the [PYTHONIOENCODING](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONIOENCODING) variable set in your environment?

**Comment by lee218llnl** [comment]
Looks like utf-8 was made default in python 3.7:

[lee218@quartz386:~]$ python-2.7.16 -c 'import locale; print(locale.getpreferredencoding())'
ANSI_X3.4-1968

[lee218@quartz386:~]$ python3-3.5.1 -c 'import locale; print(locale.getpreferredencoding())'
ANSI_X3.4-1968

[lee218@quartz386:~]$ python3-3.6.4 -c 'import locale; print(locale.getpreferredencoding())'
ANSI_X3.4-1968

[lee218@quartz386:~]$ python3-3.7.2 -c 'import locale; print(locale.getpreferredencoding())'
UTF-8

**Comment by alalazo** [comment]
I think that's a setting of your environment:
```console
$ python3.6 -c 'import locale; print(locale.getpreferredencoding())'
UTF-8
```
Will try to check what can influence this besides the environment variable above.

**Comment by lee218llnl** [comment]
Some relevant Python PEPs:

https://www.python.org/dev/peps/pep-0538/

https://www.python.org/dev/peps/pep-0540/

There was a change between 3.6 and 3.7. I'm not sure what it keys on in 3.6 that causes differences for you and me @alalazo

**Comment by alalazo** [comment]
I don't have a full grasp of what happens yet. Are the different Python versions available through module files? If so, is there any change that these module do to any `LC_*` variable?

**Comment by lee218llnl** [comment]
All of these LC Python versions were built using Spack. The executables are run without any environment modifications (i.e., no wrappers or modules). I think based on the PEPs, that with Python 3.7, they changed the way the default encoding is selected, using UTF-8 and ignoring the locale encoding.

**Comment by tgamblin** [comment]
Fixed by #13490
